include/dlaf/communication/kernels/reduce.h

```diff
@@ -37,127 +34,194 @@ namespace comm {
+                       const common::internal::ContiguousBufferHolder<T>& cont_buf, MPI_Request* req) {
   auto msg = comm::make_message(cont_buf.descriptor);
   auto& comm = pcomm.ref();
 
   DLAF_MPI_CALL(MPI_Ireduce(MPI_IN_PLACE, msg.data(), msg.count(), msg.mpi_type(), reduce_op,
                             comm.rank(), comm, req));
-  return cont_buf;
 }
 
-DLAF_MAKE_CALLABLE_OBJECT(reduceRecvInPlace);
-
-template <class T, Device D>
-auto reduceSend(comm::IndexT_MPI rank_root, common::PromiseGuard<comm::Communicator> pcomm,
-                MPI_Op reduce_op, common::internal::ContiguousBufferHolder<const T> cont_buf,
-                const matrix::Tile<const T, D>&, MPI_Request* req) {
+template <class T>
+void reduceSend(comm::IndexT_MPI rank_root, const common::PromiseGuard<comm::Communicator>& pcomm,
+                MPI_Op reduce_op, const common::internal::ContiguousBufferHolder<const T>& cont_buf,
+                MPI_Request* req) {
   auto msg = comm::make_message(cont_buf.descriptor);
   auto& comm = pcomm.ref();
 
   DLAF_MPI_CALL(
       MPI_Ireduce(msg.data(), nullptr, msg.count(), msg.mpi_type(), reduce_op, rank_root, comm, req));
 }
 
-DLAF_MAKE_CALLABLE_OBJECT(reduceSend);
-
-}
-
-template <class T, Device D>
-void scheduleReduceRecvInPlace(const comm::Executor& ex,
-                               pika::future<common::PromiseGuard<comm::Communicator>> pcomm,
-                               MPI_Op reduce_op, pika::future<matrix::Tile<T, D>> tile) {
+// TODO SenderElementType of a reference_wrapper does not work
+template <class T, class Sender>
+auto senderReduceRecvInPlace(const comm::Executor& ex,
+                             pika::future<common::PromiseGuard<comm::Communicator>> pcomm,
+                             MPI_Op reduce_op, Sender&& tile) {
   // Note:
   //
-  // GPU  -> duplicateIfNeeded ---------------------> cCPU -> MPI -------------> copyIfNeeded --> GPU
-  // CPU  ----------------------> makeItContiguous -> cCPU -> MPI -> copyBack ------------------> CPU
-  // cCPU ------------------------------------------> cCPU -> MPI ------------------------------> cCPU
+  // CPU  ---> makeItContiguous ---> cCPU -> MPI ---> copyBack ---> CPU
+  // cCPU -------------------------> cCPU -> MPI -----------------> cCPU
   //
   // where: cCPU = contiguous CPU
 
-  using pika::dataflow;
-  using pika::unwrapping;
-
-  using common::internal::copyBack_o;
-  using common::internal::makeItContiguous_o;
-  using internal::reduceRecvInPlace_o;
-  using matrix::Tile;
-  using matrix::duplicateIfNeeded;
-  using matrix::getUnwrapReturnValue;
-  using matrix::unwrapExtendTiles;
-
-  const auto& ex_copy = getHpExecutor<Backend::MC>();
+  using namespace pika::execution::experimental;
+
+  using dlaf::matrix::Tile;
+  using dlaf::internal::Policy;
+  using dlaf::internal::transform;
+  using dlaf::internal::whenAllLift;
+
+  return std::forward<Sender>(tile) | transfer(thread_pool_scheduler{}) |
```

Comment:
I *think* this is ok since this will most likely always be called from the default pool, but note that `thread_pool_scheduler` will use the pool of the calling thread if on a pika thread and otherwise the default pool. It might be worth being explicit and saying `thread_pool_scheduler{&pika::resource::get_thread_pool("default")}`, or you might want to grab this file from the CUDA scheduler PR: https://github.com/eth-cscs/DLA-Future/blob/6d4f8afc0312a4a27c6f466dea48a514ad386f06/include/dlaf/schedulers.h and use `transfer(getBackendScheduler<Backend::MC>())`.

Reply:
Or, in fact, just use `transform(Policy<Backend::MC>(), ...)`.

Reply:
Good point! I was not exactly sure of what I was doing, but my will is to run the task on the CPU, so I think your last comment is the way to go. Thanks!

Reply:
> Or, in fact, just use `transform(Policy<Backend::MC>(), ...)`.

As @albestro pointed out, `transform` doesn't actually fit here so ignore that suggestion.

Reply:
FIY @rasolca @teonnik 

As discussed, we opted for fixing it (see 4bedad4c824ffb097c2afc5652711cfd0a30b1e7) with the `getBackendScheduler` utility that will be released together with the CUDA scheduler work in a future PR.

For the moment, [include/dlaf/schedulers.h](https://github.com/eth-cscs/DLA-Future/blob/c28f3a8684276881d32f08b3517cfefbb98d2f16/include/dlaf/schedulers.h) file which implements the utility has been cherry-picked, and the part about the GPU Backend has been removed.